Enable :9093/config endpoint to get status of runtime config resources

mandarjog opened the pull request:
This PR adds `/config/` endpoint to Mixer to query mixer current runtime state for config resources.
It addresses lack of problem determination tools for mixer. If a rule or handler is not taking affect, an operator can use this endpoint to query mixer.

example when an error had occurred.
```
mjog-macbookpro:~ mjog$ curl http://localhost:9093/config/istio-config-default/rule/promtcp  | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   686  100   686    0     0   112k      0 --:--:-- --:--:-- --:--:--  133k
{
  "metadata": {
    "name": "promtcp",
    "namespace": "istio-config-default",
    "labels": {
      "istio-protocol": "tcp"
    },
    "revision": ""
  },
  "spec": {
    "actions": [
      {
        "handler": "handler.prometheus.istio-config-default",
        "instances": [
          "tcpbytesent.metric.istio-config-default",
          "tcpbytereceived.metric.istio-config-default"
        ]
      }
    ]
  },
  "status": {
    "code": "error",
    "errors": [
      "Filtering action from rule istio-config-default/promtcp. Handler handler.prometheus.istio-config-default could not be initialized due to cannot infer type information from params in instance 'requestsize.metric.istio-config-default': OR($request.size, \"abcd\") arg 2 (\"abcd\") typeError got STRING, expected INT64.",
      "Purging rule promtcp with no actions"
    ]
  }
}

```

```release-note
A config endpoint is available on the mixer self monitoring port. 
curl mixer:9093/config/$full_name_of_config_resource

If resource is not present, returns 404
If resource is present, resource.status field contains status.
If status.code == "error", status.errors contains a list of textual errors.

status codes: [ "active", "pendingUpdate", "pendingDelete", "error" ]
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/istio/mixer/1325)
<!-- Reviewable:end -->

mandarjog committed:
Add :9093/config endpoint to get status of config resources

istio-testing commented:
@mandarjog: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/mixer-presubmit.sh | 8e7c4589d5761d02cdc8a4ae28f1e43ab41202c7 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_mixer/1325/mixer-presubmit/995/) | `/test mixer-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

mandarjog committed:
fix linter errors

jmuk commented on cmd/server/cmd/server.go:
```diff
@@ -408,17 +416,39 @@ func setupServer(sa *serverArgs, info map[string]template.Info, adapters []adptr
 		fatalf("Unable to listen on socket: %v", err)
 	}
 
+	sm := http.NewServeMux()
```
Please keep existing `http.Handle` things and not defining servemux; actually this implicitly enables net/http/pprof (https://golang.org/pkg/net/http/pprof/).

mandarjog replied:
Ok fixed it.

rshriram commented:
Is this the last error for this config? or is this a validation error when user submits the config ? I guess the former, since we handle validation errors in line.
If this is the last error for this config, is it just one last error or N such errors?

jmuk commented on cmd/server/cmd/server.go:
```diff
@@ -408,17 +416,39 @@ func setupServer(sa *serverArgs, info map[string]template.Info, adapters []adptr
+			out.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		out.Header().Add("content-type", "application/json")
```
can we also allow yaml as for output?  I feel like it's more visible and debuggable for human readers.

Maybe it can be dispatched `Accept` request header.

mandarjog replied:
One can pipe the json output thru `jq` to make it more readable.

jmuk replied:
It's okay to be JSON by default. I'm saying to allow yaml when `Accept: application/yaml` header is specified.

jmuk commented on pkg/config/store/store2.go:
```diff
@@ -62,8 +62,31 @@ type BackEndResource struct {
-	Spec     proto.Message
+	Metadata ResourceMeta  `json:"metadata"`
+	Spec     proto.Message `json:"spec"`
+	Status   Status        `json:"status"`
```
Does this need to belong to Resource?  As far as I see, the status is handled separately as `statusMap` in `pkg/runtime`, so `Resource` doesn't have to hold the status as a field?

mandarjog replied:
This is somewhat consistent with k8s, where `spec`, `metadata` and `status` are all at the same level, and status is considered a sub object of the resource.

kyessenov commented:
I think you need to add more structure to Status field. Maybe make it a state machine with READY state, and then stuff errors into events instead?

mandarjog commented:
@rshriram This is independent of the validation path via webhooks on ingest. 
Mixer watches configuration and validates a snapshot of the config at a time. It is ok for the snapshot to be inconsistent. Errors are recorded during validation and associated with the config resources. 

So those errors are *all* the errors associated with a particular resource.

mandarjog committed:
use default http mux

mandarjog commented:
@kyessenov It does make sense to have conditions/ events etc. However we can evolve the status schema in next several weeks. This gets in something that is already useful.

kyessenov commented:
Should we wait until it's ready and we promote Status to an API object then?

douglas-reid commented:
I'm somewhat concerned that this is a last second addon. It'd be great if we had a full design and API review... and maybe Istio-wide consistency (Pilot).

Do we think this is OK as a hidden feature (that we don't advertise for the time being)?

Should we at least consider moving it to something like `/config/v1alpha1/resources` or something?  Or should it maybe live off the `/debug` subtree?

douglas-reid commented on pkg/config/store/store2.go:
```diff
@@ -62,8 +62,31 @@ type BackEndResource struct {
+// Status is the current status of a resource.
+type Status struct {
+	// Code for this status.
+	Code Code `json:"code"`
```
this is potentially confusing, with status.code on grpc reflecting a few well-known status codes (which are different than just strings).

mandarjog replied:
Changed names to be like `PodStatus`

douglas-reid reviewed (commented):
Can we get this another way, that doesn't expose a new endpoint?

Maybe a prom metric that looks like:

```
name: config_status
labels:
- resource
- status
- message
```

mandarjog committed:
Relocate to /debug/config/resources/

Rename Status to ResourceStatus similar to k8s PodStatus.
Rename Status.Code to ResourceStatus.Phase like PodStatus.

mandarjog commented:
Encoding status subresource as metrics does not seem right.

I changed the endpoint to `/debug/config/resources/`

Status as subresource is a standard k8s concept, we are adding it our resources.
Every K8s resource implements `/status` in a different way. `PhasePPP` is a common pattern.
We can have detailed discussion about what goes into our status, however what Is already there is useful and not far off from some other k8s statuses. 

It is fairly easy to replace current ResourceStatus with whatever we evolve it to.

geeknoid commented:
Should we punt on this PR and address the functionality in the broader context of #1495.

geeknoid commented:
This repo is no longer accepting PRs. Please resubmit this change to the istio/istio repo.

Thanks.
